compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/StagedReductionStrategy.cpp

```diff
@@ -161,15 +161,25 @@ static void buildStagedReductionStrategyThreadLevel(
   // y only will trigger the insertion of an `scf.if (threadIdx.x == 0)`
   // predicate after `scf.foreach_thread` is lowered.
   // This predicate allows further vector distribution to kick in.
+  Value root = blockCombinerOpH;
+  SmallVector<Value> opsToFuse = {gridFillH};
+  // If we have a unit dim after the reduction that doesn't broadcast fuse it
+  // wuth the reduction.
+  if (strategy.captures.maybeTrailingRank ==
```

Comment:
I am not 100% clear what this condition does, can you elaborate a bit?
Seems quite brittle to me.

Reply:
Please let me know what's the right way to do this.

Reply:
Ok, I think this is fine but requires more doc as it is quite connected to the shape of the IR we have matched and transformed up to now.

I.e.
```
By the properties matching, we know the optional trailing op takes the result of the reduction as an input argument.
It necessarily follows that maybeTrailingRank >= reductionRank - 1.
When maybeTrailingRank == reductionRank - 1, by the properties of the transformations we have applied until now, we know that the elementwise is a simple scalar operation and it can be fused in the producing reduction without
creating recomputations.

// TODO: Some `transform.assert` op that the shape of the op is indeed 1s only as a safety measure.
// TODO: More composable transform strategy parts require more matching after part of the strategy has been applied. See the discussion in #11951 for more context.
```

We may be reaching a limit of the "action at a distance / AoT staging / fusion of strategies" of transformations here: it would seem reasonable to try and separate this into a new build function that would get its own matcher.
Unfortunately this cannot just be done immediately.
The key issue is that we would need to match after transformations have been applied and we would need information such as "after transformation, this linalg op really operates on a single element". 
This would require either:
  1. multi-staged transform with C++ matchers / captures. Not sure we want to go there, at least not now
  2. adding some new transform.match that would catch all shape 
  3. starting to flesh out the matching part of the transform dialect, which is prob a few good months of work.

I strongly suspect that @ftynse prefers the 2. approach, I am not sure what his position is on 1.

In any case, this is simple enough that this is still manageable but we can't go much further and maintain sanity.

Note that the discussion above is very much related to "DFS strategies vs BFS passes" and "deeper more complex but predicatable rules" vs "simple blanket rules whose composition may be more surprising".

I expect we'll iterate a lot between the 2 in 2023 until we find the right balance: we're in the middle of transformation-first IR design.

Reply:
I generally prefer to avoid complex logic in C++ to generate schedules. This splits the optimization strategy between the logic in C++ and the logic in the dialect, which makes is more complicated and negates the benefits of the dialect being declarative. That being said, simple conditions like here don't necessarily qualify as complex logic for me.

It's not very clear to me which transformations happen until this point, but if we know the relation between ranks after those transformations rather than at the beginning, we can easily plug a simple `match_callback failures(suppress) "rank_differs_by_n"` that applies to the two handles (`maybeTiledTrailingH`, `blockCombinerOpH`) and checks the relation. If it holds, the match succeeds and returns the operations themselves, otherwise it fails and returns empty lists. Then combine the results of the match as
`%root = take_first %match_callback#0, %block_combiner_op` and `%ops_to_fuse = merge_handles %grid_fill, %match_callback#1` and voilà, no need for conditions.

This is arguably more complicated than having a condition here, but not as complex as any of the points proposed above. I won't insist though.

Reply:
> we can easily plug a simple `match_callback failures(suppress) "rank_differs_by_n"` that applies to the two handles (`maybeTiledTrailingH`, `blockCombinerOpH`) and checks the relation.

> This is arguably more complicated than having a condition here, but not as complex as any of the points proposed above.

What you spell out here seems quite similar to point 2. above with the slight difference that it has an extra C++ match callback instead of a first class `features to transform.match` ?

If that is considered good enough for the foreseeable future, this is fine with me.
Note however that this is still C++ (but structured properly) and still keeps things fused.

Maybe 2. + factoring out the part of the strategy builder to explicitly handle a generic "fuse reduction into elementwise case" is the way to go here ?

All this can be left as followup PRs, not a blocker for making progress as I think this is still manageable.

Reply:
`transform.structured.match` doesn't let one test relationships, unless you go through the ugly double-sequence-split_handle trick.

Reply:
Ah yes, I had paged that out indeed, not a slight difference at all.
Thanks!